File: Cocoa/AppFrameworks/Foundation/Calendar/YJNSCalendar.m
//
//  YJNSCalendar.m
//  YJFoundation
//
//  HomePage:https://github.com/937447974/YJCocoa
//  YJ技术支持群:557445088
//
//  Created by 阳君 on 2017/1/5.
//  Copyright © 2017年 YJCocoa. All rights reserved.
//

#import "YJNSCalendar.h"

@interface YJNSCalendar() {
    NSInteger _era;
    NSInteger _year;
    NSInteger _month;
    NSInteger _day;
    NSInteger _hour;
    NSInteger _minute;
    NSInteger _second;
    NSInteger _nanosecond;
    NSInteger _weekday;
    NSInteger _weekdayOrdinal;
    NSInteger _quarter;
    NSInteger _weekOfMonth;
    NSInteger _weekOfYear;
    NSInteger _yearForWeekOfYear;
}

@property (nonatomic, strong) NSDate *date;
@property (nonatomic, strong) NSCalendar *calendar;

@end

@implementation YJNSCalendar

+ (YJNSCalendar *)currentCalendar {
    return [[YJNSCalendar alloc] initWithDate:NSDate.date calendar:NSCalendar.currentCalendar];
}

- (instancetype)initWithDate:(NSDate *)date calendar:(NSCalendar *)calendar {
    self = [super init];
    if (self) {
        _date = date;
        _calendar = calendar;
    }
    return self;
}

- (void)initEra_Year_Month_Day {
    [self.calendar getEra:&_era year:&_year month:&_month day:&_day fromDate:self.date];
}

- (void)initHour_Minute_Second_Nanosecond {
    [self.calendar getHour:&_hour minute:&_minute second:&_second nanosecond:&_nanosecond fromDate:self.date];
}

- (void)initEra_YearForWeekOfYear_WeekOfYear_Weekday {
    [self.calendar getEra:&_era yearForWeekOfYear:&_yearForWeekOfYear weekOfYear:&_weekOfYear weekday:&_weekday fromDate:self.date];
}

#pragma mark - getter
- (NSInteger)era {
    if (_era == 0) [self initEra_Year_Month_Day];
    return _era;
}

- (NSInteger)year {
    if (_year == 0) [self initEra_Year_Month_Day];
    return _year;
}

- (NSInteger)month {
    if (_month == 0) [self initEra_Year_Month_Day];
    return _month;
}

- (NSInteger)day {
    if (_day == 0) [self initEra_Year_Month_Day];
    return _day;
}

#pragma mark
- (NSInteger)hour {
    if (_hour == 0) {
        _hour = [self.calendar component:NSCalendarUnitHour fromDate:self.date];
    }
    return _hour;
}

- (NSInteger)minute {
    if (_minute == 0) [self initHour_Minute_Second_Nanosecond];
    return _minute;
}

- (NSInteger)second {
    if (_second == 0) [self initHour_Minute_Second_Nanosecond];
    return _second;
}

- (NSInteger)nanosecond {
    if (_nanosecond == 0) [self initHour_Minute_Second_Nanosecond];
    return _nanosecond;
}

#pragma mark
- (NSInteger)weekday {
    if (_weekday == 0) [self initEra_YearForWeekOfYear_WeekOfYear_Weekday];
    return _weekday;
}

- (NSInteger)weekdayOrdinal {
    if (_weekdayOrdinal == 0) {
        _weekdayOrdinal = [self.calendar component:NSCalendarUnitWeekdayOrdinal fromDate:self.date];
    }
    return _weekdayOrdinal;
}

- (NSInteger)quarter {
    if (_quarter == 0) {
        _quarter = [self.calendar component:NSCalendarUnitQuarter fromDate:self.date];
    }
    return _quarter;
}

- (NSInteger)weekOfMonth {
    if (_weekOfMonth == 0) {
        _weekOfMonth = [self.calendar component:NSCalendarUnitWeekOfMonth fromDate:self.date];
    }
    return _weekOfMonth;
}

- (NSInteger)weekOfYear {
    if (_weekOfYear == 0) [self initEra_YearForWeekOfYear_WeekOfYear_Weekday];
    return _weekOfYear;
}

- (NSInteger)yearForWeekOfYear {
    if (_yearForWeekOfYear == 0) [self initEra_YearForWeekOfYear_WeekOfYear_Weekday];
    return _yearForWeekOfYear;
}

@end

File: YJCocoa/Classes/AppFrameworks/Foundation/Calendar/YJNSCalendar.swift
//
//  YJNSCalendar.swift
//  Pods
//
//  Created by 阳君 on 2019/5/9.
//

import Foundation

///  日历
open class YJNSCalendar: NSObject {
    
    /// 当前时间
    public class var current: YJNSCalendar {
        return YJNSCalendar(date: NSDate(timeIntervalSinceNow: 0), calendar: NSCalendar.current as NSCalendar)
    }
    
    /// NSDateComponents.era
    public var era: NSInteger {
        self.getEraYearMonthDay()
        return self._era
    }
    /// NSDateComponents.year
    public var year: NSInteger {
        self.getEraYearMonthDay()
        return self._year
    }
    /// NSDateComponents.month
    public var month: NSInteger {
        self.getEraYearMonthDay()
        return self._month
    }
    /// NSDateComponents.day
    public var day: NSInteger {
        self.getEraYearMonthDay()
        return self._day
    }
    /// NSDateComponents.hour
    public var hour: NSInteger {
        self.getHourMinuteSecondNanosecond()
        return self._hour
    }
    /// NSDateComponents.minute
    public var minute: NSInteger {
        self.getHourMinuteSecondNanosecond()
        return self._minute
    }
    /// NSDateComponents.second
    public var second: NSInteger {
        self.getHourMinuteSecondNanosecond()
        return self._second
    }
    /// NSDateComponents.nanosecond
    public var nanosecond: NSInteger {
        self.getHourMinuteSecondNanosecond()
        return self._nanosecond
    }
    /// NSDateComponents.yearForWeekOfYear
    public var yearForWeekOfYear: NSInteger {
        self.getYearForWeekOfYearWeekOfYearWeekday()
        return self._yearForWeekOfYear
    }
    /// NSDateComponents.weekOfYear
    public var weekOfYear: NSInteger {
        self.getYearForWeekOfYearWeekOfYearWeekday()
        return self._weekOfYear
    }
    /// NSDateComponents.weekday
    public var weekday: NSInteger {
        self.getYearForWeekOfYearWeekOfYearWeekday()
        return self._weekday
    }
    
    private var date: NSDate!
    private var calendar: NSCalendar!
    private var _era = 0
    private var _year = 0
    private var _month = 0
    private var _day = 0
    private var _hour = 0
    private var _minute = 0
    private var _second = 0
    private var _nanosecond = 0
    private var _weekday = 0
    private var _weekOfYear = 0
    private var _yearForWeekOfYear = 0
    
    
    /// 初始化
    public init(date: NSDate, calendar: NSCalendar) {
        super.init()
        self.date = date
        self.calendar = calendar
    }
    
    func getEraYearMonthDay() {
        if self._era > 0 || self._year > 0 || self._month > 0 || self._day > 0 {
            return
        }
        self.calendar.getEra(&_era, year: &_year, month: &_month, day: &_day, from: self.date as Date)
    }
    
    func getHourMinuteSecondNanosecond() {
        if self._hour > 0 || self._minute > 0 || self._second > 0 || self._nanosecond > 0 {
            return
        }
        self.calendar.getHour(&_hour, minute: &_minute, second: &_second, nanosecond: &_nanosecond, from: self.date as Date)
    }
    
    func getYearForWeekOfYearWeekOfYearWeekday() {
        if self._yearForWeekOfYear > 0 || self._weekOfYear > 0 || self._weekday > 0{
            return
        }
        self.calendar.getEra(nil, yearForWeekOfYear: &_yearForWeekOfYear, weekOfYear: &_weekOfYear, weekday: &_weekday, from: self.date as Date)
    }
    
}
